{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <span style=\"font-family: 'Georgia', serif;\"> Rosalind Extra Credit </span>\n",
    "### <span style=\"font-family: 'Georgia', serif;\"> Addison Yam (addisony)</span>\n",
    "### <span style=\"font-family: 'Georgia', serif;\"> BF527 - Bioinformatic Foundations Fall 2025</span>\n",
    "### <span style=\"font-family: 'Georgia', serif;\"> October 27, 2025</span>\n",
    "\n",
    "Problem 1 - Counting DNA Nucleotides\n",
    "\n",
    "Problem 2 - Transcibing DNA into RNA\n",
    "\n",
    "Problem 3 - Complementing a Strand of DNA\n",
    "\n",
    "Problem 4 - Rabbits and Recurrence Relations\n",
    "\n",
    "Problem 5 - Computing GC Content\n",
    "\n",
    "Problem 6 - Counting Point Mutations\n",
    "\n",
    "Problem 7 - Mendel's First Law\n",
    "\n",
    "Problem 8 - Translating RNA into Protein\n",
    "\n",
    "Problem 9 - Finding a Motif in DNA\t\n",
    "\n",
    "Problem 10 - "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "204 240 223 219\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Counting DNA Nucleotides\n",
    "Link: https://rosalind.info/problems/dna/\n",
    "\n",
    "Problem:\n",
    "A string is simply an ordered collection of symbols selected from some alphabet and formed into a word; \n",
    "the length of a string is the number of symbols that it contains.\n",
    "\n",
    "An example of a length 21 DNA string (whose alphabet contains the symbols 'A', 'C', 'G', and 'T') is \"ATGCTTCAGAAAGGTCTTACG.\"\n",
    "\n",
    "Given: A DNA string s\n",
    " of length at most 1000 nt.\n",
    "\n",
    "Return: Four integers (separated by spaces) counting the respective number of times that the symbols 'A', 'C', 'G', and 'T' occur in s.\n",
    "\"\"\"\n",
    "def countbases(sequence):\n",
    "    Acount=sequence.count(\"A\")\n",
    "    Gcount=sequence.count(\"G\")\n",
    "    Ccount=sequence.count(\"C\")\n",
    "    Tcount=sequence.count(\"T\")\n",
    "    return(Acount, Ccount, Gcount, Tcount)\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_dna.txt\"\n",
    "    with open(filename, 'r') as file:\n",
    "        filecontents = file.read()\n",
    "    Acount, Ccount, Gcount, Tcount = countbases(filecontents)\n",
    "    print(Acount, Ccount, Gcount, Tcount)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"countingDNA.png\" alt=\"Problem 1 Completion\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAAGUGCACAUAUCCUCGACUCGUCGUUUGAGUGAGGUGACCCACUCACUUGAAAGGAGAAGAAAGAUGGUCGAUGCCCUGUCUAAUACUCCUUCCGUCUGGGUGUGCGCCAUUUCGCAGCAAGGUAAUCCUGCAGCACACGGGAAACACUGCCGAAAGCUUGCCCUGAAAUACUCUUCGUGCUGCCGAUAAGUAUCGGAGAGGUCUCCUGGACGUGUAAUAUUGUCGCUAUUGGCAUUUUAUCACGGAAUGGCACUCGUCGUACUCACCCUCUCACAUUCGUGUUCAGCAAUCAGCUAUUUAUGGAGUAAACUAUAUAUGCUUAUGCUGUGUUCACGGCGCAUUUUCCUUCAGCUUCUGCCGAUUGAAAUUACCGAAUAAAGGCGUCAUAUACUUCGAGGACAUGAUUGAAGGGGUUUCGUCAUGAUGGUGAAUAAGACAGCUCUGUUCGAGAUAGCUCAAACAACGAGCGACGUCUAGAGACCUCGAAGGUAGGCCCGUGGAGUUCGCGUAGAACGAAUACGAAGAAUCGUAAGAGCGAGUCAACUUCUCGCUUCGCAAGUCAAUUUCGUUGCAGCAUUACGUUUCCGUACCGGUUACAGAACAUAACCGUGUCCGGCAGGACCCUGUCUCACAUGGUUUUAUCAUCUAGGCUGCUAGCGUGCGGCAUGCAGUCUUAAACCACCUCAGAACCUACCGUCUCCGCCUCAGCUCGACGUCACGUCCCGCUGUUGUCGGGCUAGGGUAGGGUCUCUGGUAUGAACGAUGGUACCGAAGAUCUGCGUAUACCAUGUGCCAUCCCCUCCAUAGAGGCAACGGGCAACAGACUACGAAAGGUCGCGAAGAACGGUUUAAUACACGGAACCAUCUUCUGAAGACCACAAGUGCGAUGCCACAAUCGUAUCCCGAUCAGGCCUCCGUCCUGUAGAGGUGCUUUAGAGGUCGGUUCCCUAGUGAUGGAACGGAGAAUUG\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Transcribing DNA into RNA\n",
    "Link:\n",
    "\n",
    "Problem:\n",
    "An RNA string is a string formed from the alphabet containing 'A', 'C', 'G', and 'U'.\n",
    "\n",
    "Given a DNA string t corresponding to a coding strand, its transcribed RNA string u \n",
    "is formed by replacing all occurrences of 'T' in t with 'U' in u.\n",
    "\n",
    "Given: A DNA string t having length at most 1000 nt.\n",
    "\n",
    "Return: The transcribed RNA string of t.\n",
    "\n",
    "Sample Dataset\n",
    "GATGGAACTTGACTACGTAAATT\n",
    "Sample Output\n",
    "GAUGGAACUUGACUACGUAAAUU\n",
    "\"\"\"\n",
    "def transcribe(sequence):\n",
    "    RNAseq = \"\"\n",
    "    for i in range(len(sequence)):\n",
    "        if sequence[i]==\"T\":\n",
    "            RNAseq+=\"U\"\n",
    "        else:\n",
    "            RNAseq+=sequence[i]\n",
    "    return(RNAseq)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_rna.txt\"\n",
    "    with open(filename, 'r') as file:\n",
    "        filecontents = file.read().strip()\n",
    "    sequence = transcribe(filecontents)\n",
    "    print(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"transcribing.png\" alt=\"Problem 2 Completion\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAGGCACTAAACTTCGTGCTCACCTCGTCGGTGACCACCATCCGAGTCTCTGAGCGCCGAGGCGTCCGCAGTTATTCCACTTGGATGTGAGCCAGTAGCAGTCGGTAGTTACGATAGACGGAGGGCTCAATAACCGAGGCATATCTTAAGGGTTCGGGTCAGGTGGCTGGCTTTACAAGAAAGCAGCTTTGAGGGAATGCGAAATAGGAAGTTTCGTCGGGTATACATGAAGAGAGGATAATTTAACGTTCGAACCAACACTTTACAAATCGGCCGAATTGCCGCAATTTGTTTAAAATGCGTTACCAAAGGGCACCTCTACCGTAAGTTGCGTTTTCGAAAGGGGGGGCCATAACGCCTCTGCCCTAATTGAGTATCCTGCTAAACGCCCCAACTTTTGGTTCTTCCAAAAAATGACACTTATTACAGTATGTCTCGTAGATAAACTCTGCGAACGGCTTGGTACGAATTAACTAATGACGGTCAGTTACGACGAAATTCTGCGCAGCGGGAGTAGATTTCTACCGATAGTTCTATGTTCCCAGCAACAGATTCGCTAACAGCTATTTAGTTAAACTCGTCTGGCTACTTCCGGGATACTGTTCTCATGGACCCGAACCTAGTCGGGGGTGGTGAGTATGACCCGTTAGGATTTTAGCTATACGATCGATAGTAGGCTCTCCACCTTAGCTATGATCTGTGAGACAGTTGGTTCAATCCGGTGTCAAGCCGAGATTGTCCCGGGATAGCTTCTGCTTGAGTCGTAACTTTCCCGACAGCGATAGCAAGCCGGCACATCATTCCAATAGTCGGCCACTTTAGGTCGGCGAGGTCCTATGCTCTAGTGGTAAACATAGGCACACTTTCATGGTTTTCTTTAGCGCGGCAGGGGTGGCCTGTGGCTCTTTTTGAAGGCATTTATTTATGGTAGCGGCGGTAC\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Complementing a Strand of DNA\n",
    "Link: https://rosalind.info/problems/revc/\n",
    "\n",
    "Problem:\n",
    "In DNA strings, symbols 'A' and 'T' are complements of each other, as are 'C' and 'G'.\n",
    "\n",
    "The reverse complement of a DNA string s is the string sc \n",
    "formed by reversing the symbols of s, then taking the complement of each symbol \n",
    "(e.g., the reverse complement of \"GTCA\" is \"TGAC\").\n",
    "\n",
    "Given: A DNA string s of length at most 1000 bp.\n",
    "\n",
    "Return: The reverse complement sc of s.\n",
    "\n",
    "Sample Dataset\n",
    "AAAACCCGGT\n",
    "Sample Output\n",
    "ACCGGGTTTT\n",
    "\"\"\"\n",
    "def complementrary(sequence):\n",
    "    complementstrand = \"\"\n",
    "    complementdict = {'A':'T', 'G':'C', 'C':'G', 'T':'A'}\n",
    "    for base in reversed(sequence):\n",
    "        complementstrand += complementdict.get(base)\n",
    "    return complementstrand\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_revc.txt\"\n",
    "    with open(filename, 'r') as file:\n",
    "        filecontents = file.read().strip()\n",
    "    print(complementrary(filecontents))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"complement.png\" alt=\"Problem 3 Completion\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "574888488199\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Rabbits and Recurrence Relations\n",
    "Link: https://rosalind.info/problems/fib/\n",
    "\n",
    "Problem:\n",
    "A sequence is an ordered collection of objects. We use a_n to represent the n-th of a sequence.\n",
    "\n",
    "A recurrence relation is a way of defininf the terms of a seuqnece with \n",
    "respect to the values of previous terms.\n",
    "\n",
    "When finding the n-th term of a sequence defined by a recurrence relation, we can simply use the recurrence relation \n",
    "to generate terms for progressively larger values of n. This problem introduces us to the computational technique of dynamic programming, \n",
    "which successively builds up solutions by using the answers to smaller cases.\n",
    "\n",
    "Input: (n, k) where n <= 40, k <=5\n",
    "\n",
    "Output: Fn = Fn-1 + Fn-2 * 3\n",
    "Total number of rabbit pairs present after n months, where every pair of reproduction age produces a litter of k raabits\n",
    "\n",
    "Sample Dataset:\n",
    "5 3\n",
    "Sample Output\n",
    "19\n",
    "\n",
    "F1 = 1\n",
    "F2 = 1\n",
    "F3 = F2 + F1 * 3 = 1 + 1 * 3 = 4\n",
    "F4 = F3 + F2 * 3 = 4 + 1 * 3 = 7\n",
    "F5 = F4 + F3 * 3 = 7 + 4 * 3 = 19\n",
    "F6 = F5 + F4 * 3 = 19 + 7 * 3 = 40\n",
    "\"\"\"\n",
    "def computefib(n, k):\n",
    "    if n==1 or n==2: #this should always be the same regardless of k\n",
    "        return 1\n",
    "    Fib = [0] * (int(n)+1) #create an array of appropriate size\n",
    "    Fib[1] = 1\n",
    "    Fib[2] = 1\n",
    "    \n",
    "    for i in range(3, int(n)+1): #initializes the array with current and previous values\n",
    "        previous = Fib[i-1]\n",
    "        offspring = Fib[i-2] * k\n",
    "        Fib[i] = previous + offspring\n",
    "    return Fib[int(n)]\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_fib.txt\"\n",
    "    with open(filename, 'r') as file:\n",
    "        line = file.read().strip()\n",
    "    parts = line.split()\n",
    "    n = int(parts[0])\n",
    "    k = int(parts[1])\n",
    "    print(computefib(n,k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"rabbits.png\" alt=\"Problem 4 Completion\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_0389\n",
      "52.85714285714286\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Computing GC Content\n",
    "Link: https://rosalind.info/problems/gc/\n",
    "\n",
    "Problem:\n",
    "The GC-content of a DNA string is given by the percentage of symbols in the string that are 'C' or 'G'. \n",
    "For example, the GC-content of \"AGCTATAG\" is 37.5%. Note that the reverse complement of any DNA string has the same GC-content.\n",
    "\n",
    "DNA strings must be labeled when they are consolidated into a database. A commonly used method of string labeling is called FASTA format. \n",
    "In this format, the string is introduced by a line that begins with '>', followed by some labeling information. Subsequent lines contain the string itself; \n",
    "the first line to begin with '>' indicates the label of the next string.\n",
    "\n",
    "In Rosalind's implementation, a string in FASTA format will be labeled by the ID \"Rosalind_xxxx\", where \"xxxx\"denotes a four-digit code between 0000 and 9999.\n",
    "\n",
    "Given: At most 10 DNA strings in FASTA format (of length at most 1 kbp each).\n",
    "\n",
    "Return: The ID of the string having the highest GC-content, followed by the GC-content of that string. Rosalind allows for a default error of 0.001 in all \n",
    "decimal answers unless otherwise stated; please see the note on absolute error below.\n",
    "\n",
    "\n",
    "Sample Dataset:\n",
    ">Rosalind_6404\n",
    "CCTGCGGAAGATCGGCACTAGAATAGCCAGAACCGTTTCTCTGAGGCTTCCGGCCTTCCC\n",
    "TCCCACTAATAATTCTGAGG\n",
    ">Rosalind_5959\n",
    "CCATCGGTAGCGCATCCTTAGTCCAATTAAGTCCCTATCCAGGCGCTCCGCCGAAGGTCT\n",
    "ATATCCATTTGTCAGCAGACACGC\n",
    ">Rosalind_0808\n",
    "CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGAC\n",
    "TGGGAACCTGCGGGCAGTAGGTGGAAT\n",
    "\n",
    "Sample Output\n",
    "Rosalind_0808\n",
    "60.919540\n",
    "\"\"\"\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_gc.txt\"\n",
    "    gc_content_dict = {}\n",
    "    current_sequence_id = None\n",
    "    current_sequence_length = 0\n",
    "    current_gc_count = 0\n",
    "\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                # New sequence record\n",
    "                if current_sequence_id is not None and current_sequence_length > 0:\n",
    "                    # Calculate and store GC content for the previous sequence\n",
    "                    gc_content = (current_gc_count / current_sequence_length) * 100\n",
    "                    gc_content_dict[current_sequence_id] = gc_content\n",
    "\n",
    "                current_sequence_id = line[1:]  # Remove '>'\n",
    "                current_sequence_length = 0\n",
    "                current_gc_count = 0\n",
    "            else:\n",
    "                # Sequence line\n",
    "                current_sequence_length += len(line)\n",
    "                for char in line:\n",
    "                    if char.upper() in ('G', 'C'):\n",
    "                        current_gc_count += 1\n",
    "\n",
    "        # Process the last sequence in the file\n",
    "        if current_sequence_id is not None and current_sequence_length > 0:\n",
    "            gc_content = (current_gc_count / current_sequence_length) * 100\n",
    "            gc_content_dict[current_sequence_id] = gc_content\n",
    "    largestGCcontent = max(gc_content_dict, key=gc_content_dict.get)\n",
    "    print(largestGCcontent)\n",
    "    print(gc_content_dict[largestGCcontent])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"ComputingGC.png\" alt=\"Problem 5 Completion\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "448\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Counting Point Mutations\n",
    "Link: https://rosalind.info/problems/hamm/\n",
    "\n",
    "Problem:\n",
    "Given two strings s and t of equal length, the Hamming distance between s and t, \n",
    "denoted dH(s,t), is the number of corresponding symbols that differ in s and t.\n",
    "\n",
    "Given: Two DNA strings s and t of equal length (not exceeding 1 kbp).\n",
    "Return: The Hamming distance dH(s,t).\n",
    "\n",
    "Sample Dataset:\n",
    "GAGCCTACTAACGGGAT\n",
    "CATCGTAATGACGGCCT\n",
    "\n",
    "Sample Output\n",
    "7\n",
    "\"\"\"\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_hamm.txt\"\n",
    "    s_seq, t_seq = \"\",\"\"\n",
    "\n",
    "    with open(filename, 'r') as f:\n",
    "        s_seq = f.readline().strip()\n",
    "        t_seq = f.readline().strip()\n",
    "\n",
    "    distance = 0\n",
    "    for i in range(len(s_seq)):\n",
    "        if s_seq[i] != t_seq[i]:\n",
    "            distance+=1\n",
    "    print(distance)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"CountingPointMutations.png\" alt=\"Problem 6 Completion\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7393196202531644\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Mendel's First Law\n",
    "Link: https://rosalind.info/problems/iprb/\n",
    "\n",
    "Problem:\n",
    "An event is a collection of outcomes. Because outcomes are distinct, the probability of an event\n",
    "can be written as the sum of the probabilities of it constituent outcomes.\n",
    "\n",
    "Given: Three positive integers k, m, and n, representing a population containing k+m+n organisms: k individuals \n",
    "are homozygous dominant for a factor, m are heterozygous, and n are homozygous recessive.\n",
    "\n",
    "Return: The probability that two randomly selected mating organisms will produce an individual possessing a \n",
    "dominant allele (and thus displaying the dominant phenotype). Assume that any two organisms can mate.\n",
    "\n",
    "Sample Dataset:\n",
    "2 2 2\n",
    "\n",
    "Sample Output:\n",
    "0.78333\n",
    "\n",
    "My thought process:\n",
    "\n",
    "For example: six genotypes: two DD, two Dd, two dd\n",
    "P(X=DD , Y=DD)=(1/3)(1/5)=(1/15) ~ 100% produce dominant\n",
    "P(X=DD , Y=Dd)=(1/3)(2/5)=(2/15) ~ 100% produce dominant\n",
    "P(X=DD , Y=dd)=(1/3)(2/5)=(2/15) ~ 100% produce dominant\n",
    "\n",
    "P(X=Dd , Y=DD)=(1/3)(2/5)=(2/15) ~ 100% produce dominant\n",
    "P(X=Dd , Y=Dd)=(1/3)(1/5)=(1/15) ~ 75% produce dominant\n",
    "P(X=Dd , Y=dd)=(1/3)(2/5)=(2/15) ~ 50% produce dominant\n",
    "\n",
    "P(X=dd , Y=dd)=(1/3)(1/5)=(1/15) ~ 0% produce dominant\n",
    "P(X=dd , Y=Dd)=(1/3)(2/5)=(2/15) ~ 50% produce dominant\n",
    "P(X=dd , Y=DD)=(1/3)(2/5)=(2/15) ~ 100% produce dominant\n",
    "\n",
    "P(dominant)=    P(X=DD) + \n",
    "                P(X=Dd , Y=DD) + P(X=Dd , Y=Dd)*75% + P(X=Dd , Y=dd)*50% + \n",
    "                P(X=dd , Y=dd)*0% + P(X=dd , Y=Dd)*75% + P(X=dd , Y=DD)*100%\n",
    "\"\"\"\n",
    "import random\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_iprb.txt\"\n",
    "    with open(filename, 'r') as f:\n",
    "        k, m, n = f.readline().strip().split()\n",
    "        k, m, n = int(k), int(m), int(n)\n",
    "    total = k+m+n\n",
    "\n",
    "    probability = 0.0 # keep track of the probability of a dominant phenotype\n",
    "\n",
    "    # get the probability one parent is homozygous dominant\n",
    "    prob_X_homo_dom = k/total\n",
    "\n",
    "    # get the probabilities one parent is heterozygous and its potential mates\n",
    "    prob_X_heter_Y_homo_dom = (m/total)*(k/(total-1))\n",
    "    prob_X_heter_Y_heter = (m/total)*((m-1)/(total-1))\n",
    "    prob_X_heter_Y_homo_rec = (m/total)*(n/(total-1))\n",
    "\n",
    "    # get the probabilities one parent is homozygous recessive and its potential mates\n",
    "    prob_X_homo_rec_Y_homo_dom = (n/total)*((k)/(total-1))\n",
    "    prob_X_homo_rec_Y_heter = (n/total)*((m)/(total-1))\n",
    "    prob_X_homo_rec_Y_homo_rec = (n/total)*((n-1)/(total-1))\n",
    "\n",
    "    # add the probability of one parent being homozygous dominant\n",
    "    probability += prob_X_homo_dom\n",
    "    # add the probability of one parent being heterozygous and its potential mates\n",
    "    probability += prob_X_heter_Y_homo_dom + prob_X_heter_Y_heter*(3/4)+prob_X_heter_Y_homo_rec*(1/2)\n",
    "    # add the probability of one parent being homozygous recessive and its potential mates\n",
    "    probability += prob_X_homo_rec_Y_homo_dom + prob_X_homo_rec_Y_heter*(1/2)\n",
    "    \n",
    "    print(probability)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"MendelsFirstLaw.png\" alt=\"Problem 7 Completed\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MGHESPRDYCPGVFLSINHFRNSIYAKGRVLLQLGGFQLVSVSRSLTNHYSVTFNCEAQVMVFMCLCRAPGNETNFKNWPSEPPLGAKLELDVSDIPGDPNIVNIQRFPAALLLNRCRVQRADVSGLNDDRLNRKTIRTCTTHNHNDFSCWPETIKVISLRQPPQGEETLHLCRHLLSCAPLRAFDLGPTSRHAVRCFLRKTAFLQLAREGAWDHWDLRQVYSAESYLPLLKLPCTVGPSRRLTKTECATEPNMAELISFGVGFFSLVVLTIDYMPSHHRASELDSVSDRETEVHRQQVHIGHLYLSTRINNSTDGDPALLNLSTIPTMVPRVWQKQLLSHRSSSDKHRASPPALSLRLDGTVSQNLAPRRVKMMAVGHGGSSTSPSDGHCSRLRPWKERKAAIPVLWSATGSSNVAGVSCILLSSSQSESYISKDGALVLQYLFFSDQLAPEVAPSGSKQGYEGYVAKLYYNRWSARRHCLTNQFRSPQEIGRRISLIGASPTIWLRYTRRRQVGHQALDPMLTRATLNELIRPSKKTIVAFVARLYEDTPAGLLGPAVYEFCFCCIYALTQDVPRPDRNGSGFVLELEMHLNDFDIVEVTPSERMLDEKGITFRLRECRSLARRDSHPGTFYQFGAFIRCTKSPTNFKCCPPRGRVRKTCKTSAKLGLIRIGVRSFCNCKTSNFYSRSRVWGRTAHASPWHAKRYIVPHGYYRICSRWQSGPTRKRGTIFCQLHLIRVRKPPTVHTTSASRSCSYRITFAGSAEVSNKRSSSTSSFVQVRRGDSEPVDSPIGTAYDPQVPVLVVVAAVNVSIQFNTQCYRGYHNVAHSPIPQWHFGNRLRRASSSSMRYYACNILILRFPLVILLLFRGELHSVVVPSQDSGPPRTPVWPAIGNKAWFKFRTPRPVGNHNSLQGTSLKFTQANSYHRHLWGNNRCDSYSGGLRLMDGLNGCLSPYISCVGRCGVMLTDSVSIYQLTWSNHCGHIQYYSMSGKTIKLRIVLRPLCILTRCDVTEEVIVLSDPDSTIAVPWKTERIGASAKAINTSFSEQCVYPSKVSRGQQPSGRTARSFYPCRDPDDPRVDSVFLILCSVDRLYRCRRASLRSPLALSPRSLRRECVSRMRYGTLRARLRHVCQGHTCADIQKTDPLSSPIRGIRSSLCASPVIASHGELAHRQGYGTAFRPFSRNRLVDTPAVVPLTLSAYASAYLYSLCQYPNQLHRFVLEHPGSCLAELVYGATRSVQNIVALKRPPDPDPHVEIIRISIPHGLSPNRQEYQTMVVDIPTMMSLLLHCFGFRVDLRLQRHLKMLKRCYYVRTRTPQSITVYRAGPAPPVRAGSYRRVWNWELGTVSALHLYVSCNVCSVLRGVKVFIEFTYLSLQAQNRLLLLIHLLYSPLPSSSASCIGIQVIGEQSRRRFILTYGDHSLACIIPWSVGPKFVKRSMALIADGSTPRPSLVSLCMAYRTGIRESRAVPPFVCPNVIVQMGGADQRFASGRLAFISACSCFYSVNQAHEEARHESGPRRRPGSLPPITFSDDFHNCAVWRPEAYCNTTFVGTVPPSKVGTQKEEKGHCLRGVTEIGVLYQMTESLGPDVKLALRGLIVLKSQLRACSLWTALTVPILPSGIQSARSHRWLREQRFSYPITGGRDGPKRAEQHGGKQRFTDIARENCERTKARLMFPYDDCKSDSGGNVWSIYQSDLVIEQFHNSCAYNLPRKGALYYIGYLISTKMLPLISSTCIIFQLRCPRPLLDVTTLKLDNLDLASPGTKSDQAQTACVHHREFDRRDSGYTYPVKMGISVPGCVLHESNSAVEVLYAAFLEALSKRLQSRETLHRAGLVLIKEGRLIPSVACSSDDKRGPKRSGLQLSKSTSNPSQRLTLPLALVIVDGNLTLRAVRFSIVQRRPARPIVLQASRDLRLSLPDALAPYHPQVRVGVNRTTRHVRAPLVVWGTAKYTHRIWIPVSPRESTLTAVSSVMAHYSAPIAAIKSLSLLFFNTYRLLTVWTGSYTTGAHQTFLKPDLTHQTQRKAVGIIWLWVAPKPALLSFDLLVVLSSTCHGQVLTEFCLVAMLILGDKNSNCLALIIGAQVPRSQFRMSLTASNRCLNHLARTNPRIFALKTDNHPSRMLGFGPIVRTRVLRSVSSRCFLHNLHLATLGAGPRGCVRCANGTSLSLAIPGSSSLENVQRVLTVRSILYLDCDSFVFSVFLWGCPLGYYVSLTLQQHCFKFLASKVRGRLGWTRGRENRLLSLPLRGEAPLQCDWLSRRGVRRAFIIGLELGAAGGLILVLDSLYYARSACTEFRYIIYVVLIRFTQGVENCGSSSIGYPSRPALLRSNRALKRGPTQSPYWISRYNNVTPCLAGTSSIYARRVLRNKNATDPRAEKFRRDCGTISMLLHRCNKASSIRSRPPNNAGLPHKPQMTRDHKHLLKSCPPGKLGYRAIGWLYERKPERAVQTCRLNGMLVNTPIKFLCVSIKGLTIALLSFIRVAPSKMRGGAFTTRQYPKPEASHRPLGRSEFRFLWTLIMNYNHFDCMLLNPCLRRTGLAEVEDSSCYENVSIGKDYQASGGRSRTLRSVCLCQPSTIPVRASVTGGVVSPTRRRRSLSAPYSEYGLEIMVPDQTGPRVISNLDPQNRETWMRSLDFGLSGSSLLPGAVRDVPAQFLHRARPQFRGPRVTGGQLFQRLMDVVLPDKSNGRCLYTVDSGILSVWTIQRMGRLGSNRQLLSNDDEDCWSNTHIDIVGRAVSRVLKIGYRPHLPNVGDGLPVKEALRCSNGPCGARSVAVRHVAHLVFGRVRIYRKGYSGGHGSLPASGRPAPTDSDKTSGFTLRTVSDYYFDEHSEGITISLGRERVCLCCMYPLTRDSPILHLSNNGAEPSCLFDSIMIQQVMIAQNFGKKESMSAYTILRVEAALVDATDTANQGPALPLAVSESLGMAYYKRTLILHTSMIRYVTH\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Translating RNA into Protein\n",
    "Link: https://rosalind.info/problems/prot/\n",
    "\n",
    "Problem:\n",
    "The 20 commonly occurring amino acids are abbreviated by using 20 letters from the English alphabet (all letters except \n",
    "for B, J, O, U, X, and Z). Protein strings are constructed from these 20 symbols. Henceforth, the term genetic string \n",
    "will incorporate protein strings along with DNA strings and RNA strings.\n",
    "\n",
    "The RNA codon table dictates the details regarding the encoding of specific codons into the amino acid alphabet.\n",
    "\n",
    "Given: An RNA string s corresponding to a strand of mRNA (of length at most 10 kbp).\n",
    "\n",
    "Return: The protein string encoded by s.\n",
    "\n",
    "Sample Dataset:\n",
    "AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\n",
    "\n",
    "Sample Output:\n",
    "MAMAPRTEINSTRING\n",
    "\"\"\"\n",
    "import random\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_prot.txt\"\n",
    "    with open(filename, 'r') as file:\n",
    "        RNA = file.read().strip()\n",
    "    \n",
    "    CODON_TABLE = {\n",
    "    'AUG': 'M', 'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L', 'UCU': 'S',\n",
    "    'UCC': 'S', 'UCA': 'S', 'UCG': 'S', 'UAU': 'Y', 'UAC': 'Y', 'UGU': 'C',\n",
    "    'UGC': 'C', 'UGG': 'W', 'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "    'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P', 'CAU': 'H', 'CAC': 'H',\n",
    "    'CAA': 'Q', 'CAG': 'Q', 'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'ACU': 'T', 'ACC': 'T', 'ACA': 'T',\n",
    "    'ACG': 'T', 'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K', 'AGU': 'S',\n",
    "    'AGC': 'S', 'AGA': 'R', 'AGG': 'R', 'GUU': 'V', 'GUC': 'V', 'GUA': 'V',\n",
    "    'GUG': 'V', 'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A', 'GAU': 'D',\n",
    "    'GAC': 'D', 'GAA': 'E', 'GAG': 'E', 'GGU': 'G', 'GGC': 'G', 'GGA': 'G',\n",
    "    'GGG': 'G', 'UAA': 'Stop', 'UAG': 'Stop', 'UGA': 'Stop'\n",
    "    }\n",
    "    protein = \"\"\n",
    "    \n",
    "    for i in range(0, len(RNA), 3): #gets every third index\n",
    "        codon = RNA[i:i+3] #gets the codon or every three RNA bases\n",
    "        if codon in [\"UAG\", \"UAA\", \"UGA\"]: #should stop when there is a stop codon\n",
    "            break\n",
    "        protein+=CODON_TABLE[codon] #add the amino acid to the protein\n",
    "\n",
    "    print(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src = \"TranslatingRNAintoProtein.png\" alt=\"Probelm 8 Solved\" width=\"400px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12 43 50 84 164 171 178 247 264 271 278 285 295 302 343 392 399 409 433 440 521 536 543 565 592 599 672 679 715 733 822 882\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Title: Finding a Motif in DNA\n",
    "Link: https://rosalind.info/problems/subs/\n",
    "\n",
    "Problem:\n",
    "Given two strings s and t, t is a substring of s. If t is contained as a contiguous collection of symbols in \n",
    "s (as a result, t must be no longer than s).\n",
    "\n",
    "The position of a symbol in a string is the total number of symbols found to its left, including itself (e.g., \n",
    "the positions of all occurrences of 'U' in \"AUGCUUCAGAAAGGUCUUACG\" are 2, 5, 6, 15, 17, and 18). The symbol at position i of s is denoted by s[i].\n",
    "\n",
    "A substring of s can be represented as s[j:k], where j and k represent the starting and ending positions of the substring in s; for example, if s = \"AUGCUUCAGAAAGGUCUUACG\", then s[2:5] = \"UGCU\".\n",
    "\n",
    "The location of a substring s[j:k] is its beginning position j; note that t will have multiple locations in s if it occurs more than once as a substring of s.\n",
    "\n",
    "Given: Two DNA strings s and t (each of length at most 1 kbp).\n",
    "\n",
    "Return: All locations of t as a substring of s.\n",
    "\n",
    "Sample Dataset:\n",
    "GATATATGCATATACTT\n",
    "ATAT\n",
    "\n",
    "Sample Output:\n",
    "2 4 10\n",
    "\n",
    "\"\"\"\n",
    "import random\n",
    "if __name__ == \"__main__\":\n",
    "    filename = \"data/rosalind_subs.txt\"\n",
    "    with open(filename, 'r') as file:\n",
    "        s = file.readline().strip()\n",
    "        t = file.readline().strip()\n",
    "    \n",
    "    coord = [] #list that will be returned\n",
    "    for i in range(len(s)-len(t)):  #loop through the length of s, minus the length of t since we're creating subsets\n",
    "        subset=s[i: i+len(t)] #creates subsets within the s string\n",
    "        if subset==t: #if that subset is equal to t, record its position\n",
    "            coord.append(i+1)\n",
    "\n",
    "    print(*coord)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"FindingMotifinDNA.png\" alt=\"Problem 9 Completed\" width=\"400px\">"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
